### PR TITLE
refactor(gateway): defineResolvers now properly handles value types

### PR DESCRIPTION
**Pull request by steinroe**
Hi,

I realized that the current valueTypes implementation is fundamentally flawed, for example, "top-level" value types were not handled properly. When a mutation/query/subscription returns a value type, it throws `Cannot return null for Mutation.myMutation`.

Hence, I refactored the way the `defineResolvers` function handles non-scalar types.  I added the documentation in-line. It basically uses `serviceForFieldType` and `serviceForType` to derive whether the field should be resolved from the parent, through a query / subscription / mutation or through an entity query.

#499

### CODE REVIEW

**Comment by mcollina** [comment]
Does this fix https://github.com/mercurius-js/mercurius/issues/499 completely?

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by steinroe** [comment]
The todos described refer to the recognition of value types and this PR only fixes the resolvers, so we cannot close the issue yet.

File: lib/gateway.js [hunk 1]
```diff
@@ -90,58 +90,82 @@ function defineResolvers (schema, typeToServiceMap, serviceMap, typeFieldsToServ
...
+              field.resolve = makeResolver({
+                service: service,
+                createOperation: createQueryOperation,
+                transformData: response => response.json.data[fieldName],
+                isQuery: true
+              })
+            }
+          } else if (serviceForFieldType !== null && serviceForFieldType !== serviceForType) {
+            /**
+             * If there is no service for the field type but a service for the type and it is not the same,
```
Comment (steinroe) [inline lib/gateway.js:138]: this should be `If there is a service for the field type`

Comment (mcollina) [reply lib/gateway.js:138]: can you update it?

Comment (steinroe) [reply lib/gateway.js:138]: Yes, I also saw that the docs on function level are not required anymore. Should I remove them or rather change it to reflect the new design?

Comment (mcollina) [reply lib/gateway.js:138]: Updated them to the new design

File: lib/gateway.js [hunk 2]
```diff
@@ -90,58 +90,82 @@ function defineResolvers (schema, typeToServiceMap, serviceMap, typeFieldsToServ
...
+               *    userServiceInfo: SomeReturnType!
+               * }
+               * where SomeReturnType is a value type.
+               * In these cases, we get the service from the typeFieldsToService map.
+               */
+              service = serviceMap[typeFieldsToService[`${type}-${fieldName}`]]
+            }
+            if (type.name === 'Subscription') {
+              field.subscribe = makeResolver({
+                service: service,
```
Comment (jonnydgreen) [inline lib/gateway.js:123]:
```suggestion
                service,
```

File: lib/gateway.js [hunk 3]
```diff
@@ -90,58 +90,82 @@ function defineResolvers (schema, typeToServiceMap, serviceMap, typeFieldsToServ
...
+                createOperation: createQueryOperation,
+                isQuery: true,
+                isSubscription: true
+              })
+            } else {
+              field.resolve = makeResolver({
+                service: service,
```
Comment (jonnydgreen) [inline lib/gateway.js:130]:
```suggestion
                service,
```

**Review by jonnydgreen — COMMENTED** [review]
Looks great! I really like the comments you've added 🚀

File: lib/gateway.js (hunk 1 above)
Comment (steinroe) [reply lib/gateway.js]: I updated the docs :)

**Comment by simoneb** [comment]
Just dropping a line to say I've been working right today on a similar area to make subscriptions work in gateway with scalar types https://github.com/mercurius-js/mercurius/pull/508. Not sure if the two are related but I thought I'd mention this anyway. What I found is that handling of types needs to be consistent when done through a federation gateway, or you'll end up with things which work when in a classic scenario but not when in gateway mode

**Pull request merged**